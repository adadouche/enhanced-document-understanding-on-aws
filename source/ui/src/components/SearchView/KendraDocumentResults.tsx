/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useMemo } from 'react';

import { SpaceBetween } from '@cloudscape-design/components';
import KendraHighlightedText from './KendraHighlightedText';
import KendraResultTitle from './KendraResultTitle';

export default function KendraDocumentResults({
    results,
    casesList,
    setSelectedCaseId,
    setSelectedDocumentId,
    setSelectedDocumentFileType
}: any) {
    const renderedResults = useMemo(
        () => (
            <SpaceBetween direction="vertical" size="l">
                {results.map((result: any) => {
                    const caseObject = casesList.find((obj: any) =>
                        obj.caseDocuments.some((doc: any) => doc.docId === result.DocumentId)
                    );
                    return (
                        <article key={result.Id} data-testid="kendra-result-title">
                            <KendraResultTitle
                                result={result}
                                caseObject={caseObject}
                                setSelectedCaseId={setSelectedCaseId}
                                setSelectedDocumentId={setSelectedDocumentId}
                                setSelectedDocumentFileType={setSelectedDocumentFileType}
                            />
                            <p data-testid="kendra-highlighted-text">
                                <KendraHighlightedText textWithHighlights={result.DocumentExcerpt} />
                            </p>
                        </article>
                    );
                })}
            </SpaceBetween>
        ),
        [casesList, results, setSelectedCaseId, setSelectedDocumentFileType, setSelectedDocumentId]
    );

    return <div>{renderedResults}</div>;
}
