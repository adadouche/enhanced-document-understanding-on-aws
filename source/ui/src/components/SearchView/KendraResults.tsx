/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import KendraResultPage from './KendraResultPage';

import { SpaceBetween } from '@cloudscape-design/components';
import { MIN_SEARCH_QUERY_LENGTH } from '../../utils/constants';

type KendraResultsProps = {
    results: any[];
    searchQuery: string;
    casesList: any[];
    setSelectedCaseId: Function;
    setSelectedDocumentId: Function;
    setSelectedDocumentFileType: Function;
};

export default function KendraResults(props: KendraResultsProps) {
    const isQueryLongEnough = props.searchQuery && props.searchQuery.length >= MIN_SEARCH_QUERY_LENGTH;
    if (!props.searchQuery) return null;

    return (
        <div>
            <SpaceBetween direction="vertical" size="l">
                <header>
                    <h2>Amazon Kendra Results</h2>
                </header>

                <div>
                    {isQueryLongEnough && props.results.length > 0 && (
                        <KendraResultPage
                            results={props.results}
                            casesList={props.casesList}
                            setSelectedCaseId={props.setSelectedCaseId}
                            setSelectedDocumentId={props.setSelectedDocumentId}
                            setSelectedDocumentFileType={props.setSelectedDocumentFileType}
                        />
                    )}
                    {props.results.length < 1 && <div>No results found</div>}
                </div>
            </SpaceBetween>
        </div>
    );
}
