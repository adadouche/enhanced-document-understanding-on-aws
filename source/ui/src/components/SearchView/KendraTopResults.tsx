/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useCallback, useMemo, useState } from 'react';

import KendraHighlightedText from './KendraHighlightedText';
import KendraResultTitle from './KendraResultTitle';

const MAX_TOP_ANSWER_LENGTH = 5; // temp

function getTopAnswer(text: any) {
    if (text && text.Highlights) {
        for (const highlight of text.Highlights) {
            const length = highlight.EndOffset - highlight.BeginOffset;
            if (highlight && highlight.TopAnswer && length < MAX_TOP_ANSWER_LENGTH) {
                return <h1>{text.Text.substring(highlight.BeginOffset, highlight.EndOffset)}</h1>;
            }
        }
    }

    return null;
}

export default function KendraTopResults({
    results,
    casesList,
    setSelectedCaseId,
    setSelectedDocumentId,
    setSelectedDocumentFileType
}: any) {
    const renderedResults = useMemo(
        () =>
            results?.map((result: any) => {
                const answer = result?.AdditionalAttributes?.find((attr: any) => attr.Key === 'AnswerText');

                const caseObject = casesList?.find((obj: any) =>
                    obj.caseDocuments.some((doc: any) => doc.name === result.DocumentId)
                );
                return (
                    <article key={result.id}>
                        <KendraResultTitle
                            result={result}
                            caseObject={caseObject}
                            setSelectedCaseId={setSelectedCaseId}
                            setSelectedDocumentId={setSelectedDocumentId}
                            setSelectedDocumentFileType={setSelectedDocumentFileType}
                        />
                        {getTopAnswer(answer?.TextWithHighlightsValue)}
                        <p>
                            <KendraHighlightedText textWithHighlights={answer.Value.TextWithHighlightsValue} />
                        </p>
                    </article>
                );
            }),
        [casesList, results, setSelectedCaseId, setSelectedDocumentFileType, setSelectedDocumentId]
    );

    const [expanded, setExpanded] = useState(false);

    const toggleExpanded = useCallback(() => {
        setExpanded((e) => !e);
    }, []);

    if (!results.length) return null;

    return (
        <div>
            <header>
                <h2>Amazon Kendra suggested answers</h2>
            </header>
            <div>
                {renderedResults[0]}

                {renderedResults.length > 1 ? (
                    <>
                        <div onClick={toggleExpanded}>More suggested answers ({renderedResults.length - 1})</div>
                        {expanded ? <div>{renderedResults.slice(1)}</div> : null}
                    </>
                ) : null}
            </div>
        </div>
    );
}
