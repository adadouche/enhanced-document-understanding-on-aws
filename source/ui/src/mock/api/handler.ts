// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { rest } from 'msw';

export const MOCK_CONFIG = {
    ApiEndpoint: 'https://jsonplaceholder.typicode.com/',
    AWS_REGION: 'us-east-1',
    USER_POOL_ID: 'test_user_pool_id',
    USER_POOL_CLIENT_ID: 'testClientId',
    DATA_BUCKET_NAME: 'test-bucket'
};

export const getSignedUrlResponse = {
    downloadUrl: `${MOCK_CONFIG.ApiEndpoint}fake-url`
};

const getDocumentResponse = {
    key: `${MOCK_CONFIG.ApiEndpoint}/fake-key`
};

const getCasesResponse = {
    Count: 4,
    ScannedCount: 4,
    Items: [
        {
            CASE_ID: { S: 'fake-case-id-1' },
            CASE_NAME: { S: 'fake-case-name-1' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '1' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        },
        {
            CASE_ID: { S: 'fake-case-id-2' },
            CASE_NAME: { S: 'fake-case-name-2' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '0' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        },
        {
            CASE_ID: { S: 'fake-case-id-3' },
            CASE_NAME: { S: 'fake-case-name-3' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '0' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        },
        {
            CASE_ID: { S: 'fake-case-id-4' },
            CASE_NAME: { S: 'fake-case-name-4' },
            DOCUMENT_ID: { S: '0000' },
            STATUS: { S: 'success' },
            USER_ID: { S: 'fake-user-id' },
            DOC_COUNT: { N: '0' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        }
    ]
};

const getDocumentByCaseIdResponse = {
    Count: 1,
    ScannedCount: 1,
    Items: [
        {
            BUCKET_NAME: { S: 'fake-bucket-name-1' },
            CASE_ID: { S: 'fake-case-id-1' },
            CASE_NAME: { S: 'fake-case-name-1' },
            DOCUMENT_ID: { S: 'fake-doc-id1' },
            DOCUMENT_TYPE: { S: 'passport' },
            S3_KEY: {
                S: 'fake-s3-key.jpg'
            },
            UPLOADED_FILE_EXTENSION: { S: 'jpg' },
            UPLOADED_FILE_NAME: { S: 'fake-file-name' },
            USER_ID: { S: 'fake-user-id' },
            CREATION_TIMESTAMP: { 'S': '2000-01-01T00:00:00.000Z' }
        }
    ]
};

export const handlers = [
    rest.get('/runtimeConfig.json', (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(MOCK_CONFIG)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}cases`, (req, res, ctx) => {
        let response: any = getCasesResponse;
        if (req.params) {
            response = {
                ...getCasesResponse,
                LastEvaluatedKey: true
            };
        }
        return res(ctx.status(200), ctx.text(JSON.stringify(response)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}case/:caseId`, (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(getDocumentByCaseIdResponse)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}document/:caseId/:documentId`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(getDocumentResponse)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}document/download`, (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(getSignedUrlResponse)));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}inferences/:caseId/:documentId`, (_, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify(['textract-detectText', 'entity-medical-locations'])));
    }),
    rest.get(`${MOCK_CONFIG.ApiEndpoint}inferences/:caseId/:documentId/:inferenceType`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.text(JSON.stringify({})));
    }),
    rest.post(`${MOCK_CONFIG.ApiEndpoint}document`, (_, res, ctx) => {
        return res(
            ctx.status(200),
            ctx.text(
                JSON.stringify({
                    fields: [
                        {
                            key: 'test-entry'
                        }
                    ],
                    url: 'signed-url'
                })
            )
        );
    }),
    rest.all('/signed-url', (req, res, ctx) => {
        return res(ctx.status(403));
    })
];
