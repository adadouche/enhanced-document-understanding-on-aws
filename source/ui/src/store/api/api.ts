/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { BaseQueryFn, createApi, FetchArgs, fetchBaseQuery, FetchBaseQueryError } from '@reduxjs/toolkit/query/react';
import { generateToken, getRuntimeConfig } from '../../utils/apiHelper';

export const dynamicBaseQuery: BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError> = async (
    args: any,
    api: any,
    extraOptions: any
) => {
    const runtimeConfig = await getRuntimeConfig();
    const rawBaseQuery = fetchBaseQuery({
        baseUrl: runtimeConfig.ApiEndpoint,
        prepareHeaders: async (headers: any) => {
            if (process.env.NODE_ENV !== 'test') {
                const token = await generateToken();
                headers.set('Authorization', `${token}`);
            }
            return headers;
        }
    });
    return rawBaseQuery(args, api, extraOptions);
};
export const api = createApi({
    reducerPath: 'api',
    baseQuery: dynamicBaseQuery,
    tagTypes: ['Cases'],
    endpoints: () => ({
       
    }),
    refetchOnMountOrArgChange: true
});

