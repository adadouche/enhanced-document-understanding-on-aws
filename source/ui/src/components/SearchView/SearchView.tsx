/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import {
    AppLayout,
    Button,
    Checkbox,
    Container,
    ContentLayout,
    Form,
    FormField,
    Header,
    Input,
    Multiselect,
    MultiselectProps,
    SpaceBetween,
    Spinner
} from '@cloudscape-design/components';
import { OptionDefinition } from '@cloudscape-design/components/internal/components/option/interfaces';
import { API } from 'aws-amplify';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import { API_NAME, FacetDocumentAttributeKey, MIN_SEARCH_QUERY_LENGTH } from '../../utils/constants';
import { FacetResult } from '../../utils/interfaces';
import { generateToken } from '../DocumentTable/DocumentTable';
import { mapResultsToCases } from '../makeData';
import KendraResults from './KendraResults';

type SearchViewProps = {
    searchValue: string;
    setSearchValue: Function;
    submittedSearchValue: string;
    setSubmittedSearchValue: Function;
    casesList: any[];
    setCasesList: Function;
    caseName: string;
    setSelectedCaseId: Function;
    setSelectedDocumentId: Function;
    setSelectedDocumentFileType: Function;
};
let token: string;
let results: any;

const QUERY_LONG_ENOUGH_ERROR = `Enter a search query longer than ${MIN_SEARCH_QUERY_LENGTH - 1} characters to
initiate a search.`;

export default function SearchView(props: SearchViewProps) {
    let cases: any[];
    if (props.casesList) {
        cases = [
            ...new Set(
                props.casesList.map((item) => {
                    return { 'name': item.name, 'id': item.caseId };
                })
            )
        ];
    } else {
        cases = [];
    }
    const casesOptions = cases.map((item) => {
        return {
            label: item.name,
            value: item.id
        };
    });

    const [selectedCaseOptions, setSelectedCaseOptions] = React.useState<MultiselectProps.Option[]>([]);
    const [selectedFacets, setSelectedFacets] = React.useState<{ [key: string]: string[] }>({});

    const [kendraResults, setKendraResults] = React.useState<any[]>([]);
    const [facetResults, setFacetResults] = React.useState<any[]>([]);
    const [searchValueError, setSearchValueError] = React.useState('');
    const [loading, setLoading] = React.useState(false);

    React.useEffect(() => {
        const fetchData = async () => {
            token = await generateToken();
            results = await API.get(API_NAME, 'cases', {
                headers: {
                    Authorization: token
                }
            });
            const newData = mapResultsToCases(results.Items);
            props.setCasesList(newData);
        };
        if (!props.casesList) {
            fetchData();
        }
    }, []);

    const queryKendra = async (facetsToFilterBy = selectedFacets, maintainFacetSelection = false) => {
        if (props.submittedSearchValue) {
            try {
                setLoading(true);
                token = await generateToken();
                const queryParams: any = {
                    headers: {
                        Authorization: token
                    }
                };
                if (selectedCaseOptions.length > 0) {
                    queryParams.queryStringParameters = {
                        'case_id': selectedCaseOptions.map((caseOptions) => caseOptions.value)
                    };
                }
                if (Object.keys(facetsToFilterBy).length > 0 && maintainFacetSelection) {
                    queryParams.queryStringParameters = {
                        ...queryParams.queryStringParameters,
                        ...facetsToFilterBy
                    };
                }
                const response = await API.get(
                    API_NAME,
                    `search/kendra/${encodeURIComponent(props.submittedSearchValue)}`,
                    queryParams
                );
                setKendraResults(response?.ResultItems);
                setFacetResults(response?.FacetResults);

                const uniqueFacets: Set<string> = new Set();
                response?.FacetResults.forEach((result: FacetResult) => {
                    uniqueFacets.add(result.DocumentAttributeKey);
                });
                const emptySelectedFacets: Record<string, string[]> = {};
                uniqueFacets.forEach((key: string) => {
                    emptySelectedFacets[key] = [];
                });
                if (maintainFacetSelection) {
                    for (const key of Object.keys(emptySelectedFacets)) {
                        emptySelectedFacets[key] = facetsToFilterBy[key];
                    }
                }
                setSelectedFacets(emptySelectedFacets);
                setLoading(false);
                return response;
            } catch (err) {
                console.error(err);
            }
        } else {
            setLoading(false);
        }
    };

    React.useEffect(() => {
        queryKendra();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.submittedSearchValue, selectedCaseOptions]);

    let documentsTotal;

    if (props.casesList) {
        documentsTotal = props.casesList.reduce((total, caseObject) => {
            return total + caseObject.caseDocuments.length;
        }, 0);
    } else {
        documentsTotal = 0;
    }

    const navigate = useNavigate();

    const checkForSearchValueErrors = () => {
        if (!isQueryLongEnough) {
            setSearchValueError(QUERY_LONG_ENOUGH_ERROR);
            return true;
        }
        return false;
    };

    const handleKeyDown = async (key: string) => {
        if (key === 'Enter' && !checkForSearchValueErrors() && props.submittedSearchValue !== props.searchValue) {
            setLoading(true);
            setSearchValueError('');
            props.setSubmittedSearchValue(props.searchValue);
        }
    };

    const handleSearchClick = async () => {
        console.count('search click');
        if (!checkForSearchValueErrors() && props.submittedSearchValue !== props.searchValue) {
            setLoading(true);
            setSearchValueError('');
            props.setSubmittedSearchValue(props.searchValue);
        }
    };

    const handleCheckboxClick = async (facetType: string, facetValue: string) => {
        setLoading(true);
        if (isFacetSelected(facetType, facetValue)) {
            setSelectedFacets({
                ...selectedFacets,
                [facetType]: selectedFacets[facetType]?.filter((value: string) => value !== facetValue) ?? []
            });
            queryKendra(
                {
                    ...selectedFacets,
                    [facetType]: selectedFacets[facetType]?.filter((value: string) => value !== facetValue) ?? []
                },
                true
            );
        } else {
            setSelectedFacets({
                ...selectedFacets,
                [facetType]: [...selectedFacets[facetType], facetValue]
            });
            queryKendra(
                {
                    ...selectedFacets,
                    [facetType]: [...selectedFacets[facetType], facetValue]
                },
                true
            );
        }
    };

    const isFacetSelected = (facetType: string, facetValue: string) => {
        if (Object.keys(selectedFacets).includes(facetType)) {
            return selectedFacets[facetType].includes(facetValue);
        }
        return false;
    };

    if (documentsTotal === 0) {
        return (
            <div
                style={{
                    position: 'relative',
                    minHeight: '320rem',
                    paddingBottom: '32rem'
                }}
            >
                <p
                    style={{
                        margin: '2em 0',
                        textAlign: 'center',
                        fontWeight: 'bold',
                        color: 'gray'
                    }}
                >
                    No documents found. <br />
                    <Button variant="normal" onClick={() => navigate('/uploadDocument')}>
                        Add a new Document
                    </Button>
                </p>
            </div>
        );
    }

    const isQueryLongEnough = props.searchValue && props.searchValue.length >= MIN_SEARCH_QUERY_LENGTH;

    return (
        <AppLayout
            contentType="form"
            content={
                <ContentLayout
                    header={
                        <SpaceBetween size="m">
                            <Header variant="h1">Kendra search</Header>
                        </SpaceBetween>
                    }
                    data-testid="kendra-search-contentlayout"
                >
                    <form onSubmit={(event) => event.preventDefault()}>
                        <Form
                            actions={
                                <SpaceBetween direction="horizontal" size="xs">
                                    <Button
                                        data-testid="kendra-search-button"
                                        variant="primary"
                                        onClick={() => handleSearchClick()}
                                    >
                                        Search
                                    </Button>
                                </SpaceBetween>
                            }
                            errorIconAriaLabel="Error"
                            data-testid="search-view-form"
                        >
                            <Container>
                                <SpaceBetween direction="vertical" size="l">
                                    <FormField
                                        label="Search query"
                                        i18nStrings={{ errorIconAriaLabel: 'Error' }}
                                        data-testid="search-field"
                                        errorText={searchValueError}
                                    >
                                        <Input
                                            placeholder="Search"
                                            type="search"
                                            data-testid="search-view-search"
                                            value={props.searchValue}
                                            onChange={({ detail }) => props.setSearchValue(detail.value)}
                                            onKeyDown={({ detail }) => handleKeyDown(detail.key)}
                                        />
                                    </FormField>
                                    <FormField
                                        label="Case selection"
                                        i18nStrings={{ errorIconAriaLabel: 'Error' }}
                                        data-testid="case-selection-field"
                                    >
                                        <Multiselect
                                            selectedOptions={selectedCaseOptions}
                                            onChange={({ detail }) => {
                                                setSelectedCaseOptions(detail.selectedOptions as OptionDefinition[]);
                                            }}
                                            options={casesOptions}
                                            filteringType="auto"
                                            selectedAriaLabel="Selected"
                                            data-testid="search-case-multiselect"
                                            placeholder="Choose to specify cases"
                                        />
                                    </FormField>
                                </SpaceBetween>
                            </Container>
                        </Form>
                    </form>
                    <div
                        style={{
                            position: 'relative',
                            minHeight: '320rem',
                            paddingBottom: '32rem'
                        }}
                    >
                        {loading && <Spinner size="large" />}
                        {props.submittedSearchValue && (
                            <div style={{ display: 'flex', height: '100%' }}>
                                {facetResults.length > 0 && (
                                    <div
                                        style={{
                                            width: '25%',
                                            float: 'left'
                                        }}
                                    >
                                        <h2>Filters</h2>
                                        {facetResults.map((facetResult: any) => (
                                            <div key={JSON.stringify(facetResult)}>
                                                {Object.keys(FacetDocumentAttributeKey).includes(
                                                    facetResult.DocumentAttributeKey
                                                ) && (
                                                    <div>
                                                        <h4>
                                                            {
                                                                FacetDocumentAttributeKey[
                                                                    facetResult.DocumentAttributeKey as keyof typeof FacetDocumentAttributeKey
                                                                ]
                                                            }
                                                        </h4>
                                                        {facetResult.DocumentAttributeValueCountPairs.map(
                                                            (valueCountPair: any) => (
                                                                <Checkbox
                                                                    onChange={({ detail }) =>
                                                                        handleCheckboxClick(
                                                                            facetResult.DocumentAttributeKey,
                                                                            valueCountPair.DocumentAttributeValue
                                                                                .StringValue
                                                                        )
                                                                    }
                                                                    checked={isFacetSelected(
                                                                        facetResult.DocumentAttributeKey,
                                                                        valueCountPair.DocumentAttributeValue
                                                                            .StringValue
                                                                    )}
                                                                    key={JSON.stringify(valueCountPair)}
                                                                >
                                                                    <p>
                                                                        {(facetResult.DocumentAttributeKey ===
                                                                        'file_type'
                                                                            ? valueCountPair.DocumentAttributeValue.StringValue.slice(
                                                                                  1
                                                                              )
                                                                            : valueCountPair.DocumentAttributeValue
                                                                                  .StringValue) +
                                                                            ' (' +
                                                                            valueCountPair.Count.toString() +
                                                                            ')'}
                                                                    </p>
                                                                </Checkbox>
                                                            )
                                                        )}
                                                    </div>
                                                )}
                                            </div>
                                        ))}
                                    </div>
                                )}
                                <div
                                    style={{
                                    float: 'right'
                                    }}
                                >
                                    <KendraResults
                                        results={kendraResults}
                                        searchQuery={props.submittedSearchValue}
                                        casesList={props.casesList}
                                        setSelectedCaseId={props.setSelectedCaseId}
                                        setSelectedDocumentId={props.setSelectedDocumentId}
                                        setSelectedDocumentFileType={props.setSelectedDocumentFileType}
                                    />
                                </div>
                            </div>
                        )}
                    </div>
                </ContentLayout>
            }
            headerSelector="#header"
            navigationHide
            toolsHide
            date-testid="search-view-applayout"
        />
    );
}
