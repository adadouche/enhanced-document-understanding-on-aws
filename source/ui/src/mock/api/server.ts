// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { handlers } from "./handler";
import { setupServer } from "msw/node";
export const server = setupServer(...handlers);