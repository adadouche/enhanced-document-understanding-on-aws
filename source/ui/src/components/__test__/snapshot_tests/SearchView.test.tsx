/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import { API } from '@aws-amplify/api';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import SearchView from '../../SearchView/SearchView';

const mockAPI = {
    get: jest.fn(),
    post: jest.fn()
};
jest.mock('@aws-amplify/api');
API.get = mockAPI.get;
API.post = mockAPI.post;

beforeEach(() => {
    mockAPI.post.mockReset();
});

const props = {
    searchValue: '',
    setSearchValue: jest.fn(),
    submittedSearchValue: '',
    setSubmittedSearchValue: jest.fn(),
    casesList: [],
    setCasesList: jest.fn(),
    caseName: '',
    setSelectedCaseId: jest.fn(),
    setSelectedDocumentId: jest.fn(),
    setSelectedDocumentFileType: jest.fn()
};
const propsWithCases = {
    searchValue: '',
    setSearchValue: jest.fn(),
    submittedSearchValue: '',
    setSubmittedSearchValue: jest.fn(),
    casesList: [{ name: 'case1', caseDocuments: [{ name: 'doc1' }] }],
    setCasesList: jest.fn(),
    caseName: '',
    setSelectedCaseId: jest.fn(),
    setSelectedDocumentId: jest.fn(),
    setSelectedDocumentFileType: jest.fn()
};

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

test('Snapshot test', async () => {
    const tree = renderer
        .create(
            <MemoryRouter initialEntries={['/search']}>
                <Routes>
                    <Route path="/search" element={<SearchView {...props} />} />
                </Routes>
            </MemoryRouter>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('Snapshot test with cases', async () => {
    const tree = renderer
        .create(
            <MemoryRouter initialEntries={['/search']}>
                <Routes>
                    <Route path="/search" element={<SearchView {...propsWithCases} />} />
                </Routes>
            </MemoryRouter>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
