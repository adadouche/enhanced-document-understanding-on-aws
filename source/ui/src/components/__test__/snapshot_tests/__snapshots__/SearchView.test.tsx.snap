// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
<div
  style={
    Object {
      "minHeight": "320rem",
      "paddingBottom": "32rem",
      "position": "relative",
    }
  }
>
  <p
    style={
      Object {
        "color": "gray",
        "fontWeight": "bold",
        "margin": "2em 0",
        "textAlign": "center",
      }
    }
  >
    No documents found. 
    <br />
    <Button
      onClick={[Function]}
      variant="normal"
    >
      Add a new Document
    </Button>
  </p>
</div>
`;

exports[`Snapshot test with cases 1`] = `
<AppLayout
  content={
    <ContentLayout
      data-testid="kendra-search-contentlayout"
      header={
        <SpaceBetween
          size="m"
        >
          <Header
            variant="h1"
          >
            Kendra search
          </Header>
        </SpaceBetween>
      }
    >
      <form
        onSubmit={[Function]}
      >
        <Form
          actions={
            <SpaceBetween
              direction="horizontal"
              size="xs"
            >
              <Button
                data-testid="kendra-search-button"
                onClick={[Function]}
                variant="primary"
              >
                Search
              </Button>
            </SpaceBetween>
          }
          data-testid="search-view-form"
          errorIconAriaLabel="Error"
        >
          <Container>
            <SpaceBetween
              direction="vertical"
              size="l"
            >
              <FormField
                data-testid="search-field"
                errorText=""
                i18nStrings={
                  Object {
                    "errorIconAriaLabel": "Error",
                  }
                }
                label="Search query"
              >
                <Input
                  data-testid="search-view-search"
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Search"
                  type="search"
                  value=""
                />
              </FormField>
              <FormField
                data-testid="case-selection-field"
                i18nStrings={
                  Object {
                    "errorIconAriaLabel": "Error",
                  }
                }
                label="Case selection"
              >
                <Multiselect
                  data-testid="search-case-multiselect"
                  filteringType="auto"
                  onChange={[Function]}
                  options={
                    Array [
                      Object {
                        "label": "case1",
                        "value": undefined,
                      },
                    ]
                  }
                  placeholder="Choose to specify cases"
                  selectedAriaLabel="Selected"
                  selectedOptions={Array []}
                />
              </FormField>
            </SpaceBetween>
          </Container>
        </Form>
      </form>
      <div
        style={
          Object {
            "minHeight": "320rem",
            "paddingBottom": "32rem",
            "position": "relative",
          }
        }
      >
        
      </div>
    </ContentLayout>
  }
  contentType="form"
  date-testid="search-view-applayout"
  headerSelector="#header"
  navigationHide={true}
  toolsHide={true}
/>
`;
